README.markdown
# roar-rails

_Makes using Roar's representers in your Rails app fun._

[![Build Status](https://travis-ci.org/apotonick/roar-rails.svg?branch=master)](https://travis-ci.org/apotonick/roar-rails)
[![Gem Version](https://badge.fury.io/rb/roar-rails.svg)](http://badge.fury.io/rb/roar-rails)

_roar-rails development will be discontinued in the future and we will encourage users to begin migrating to [Trailblazer](http://trailblazer.to) (and [trailblazer-rails](https://github.com/trailblazer/trailblazer-rails))._

---

Roar is a framework for parsing and rendering REST documents. For a better overview about representers please check the [roar repository](https://github.com/trailblazer/roar#roar).

Roar-rails gives you conventions and convenient access to a lot of Roar's functionality within your Rails app.

## Features

* Rendering with responders
* Parsing incoming documents
* URL helpers in representers
* Better tests
* Autoloading
* Generators

This gem works with all Rails >= 3.x.

## Prerequisites

Add it to your app's `Gemfile`.

```ruby
gem "roar-rails"
```

Note: For Rails >= 4.2, you need to add the `responders` gem, too, if you use `respond_with`. This has to be before the `roar-rails` entry in the Gemfile.

```ruby
gem "responders"
gem "roar-rails"
```

## Generators

The generator will create the representer modules in `app/representers` for you.

Here's an example.

```shell
rails g representer Band id name
```

This will create the file `app/representers/band_representer.rb` with the following content,

```ruby
  module BandRepresenter
    include Roar::JSON

    property :id
    property :name
  end
```

You can change the format (e.g. XML), and pass arbitrary options to customize the generated representer. For all available options, just run

```shell
rails g representer
```


## Rendering with #respond_with

roar-rails provides a number of baked-in rendering methods.


### Conventional Rendering

Easily render resources using representers with the built-in responder.

```ruby
class SingersController < ApplicationController
  include Roar::Rails::ControllerAdditions
  respond_to :json

  def show
    singer = Singer.find_by_id(params[:id])
    respond_with singer
  end
end
```

The representer name will be infered from the passed model class (e.g. a `Singer` instance gets the `SingerRepresenter`). If the passed model is a collection it will be extended using a representer. The representer name will be computed from the controller name (e.g. a `SingersController` uses the `SingersRepresenter`).

Need to use a representer with a different name than your model? You may always pass it in using the `:represent_with` option:

```ruby
respond_with singers, :represent_with => MusicianCollectionRepresenter
end
```

### Represents Configuration

If you don't want to use conventions or pass representers you can configure them on the class level using `::represents`. This will also call `respond_to` for you.

```ruby
class SingersController < ApplicationController
  represents :json, Musician
```
This will use the `MusicianRepresenter` for models and `MusiciansRepresenter` for representing collections.

Note that `::represents` also allows fine-tuning.

```ruby
class SingersController < ApplicationController
  represents :json, :entity => MusicianRepresenter, :collection => MusicianCollectionRepresenter
```

You might pass strings as representer names to `::represents`, they will be constantized at run-time when needed.

## Rendering with #render

In place of `#respond_with`, you can also use `#render` to serialize objects using representers.

```ruby
class SingersController < ApplicationController
  include Roar::Rails::ControllerAdditions
  include Roar::Rails::ControllerAdditions::Render

  def show
    singer = Singer.find_by_id(params[:id])
    render json: singer
  end
end
```

### Old API Support

If you don't want to write a dedicated representer for a collection of items (highly recommended, thou) but rather use a representer for each item, use the `:represent_items_with` option.

```ruby
class SingersController < ApplicationController

  def index
    singers = Musician.find(:all)
    respond_with singers, :represent_items_with => SingerRepresenter
  end
end
```


## Parsing incoming documents

In `#create` and `#update` actions it is often necessary to parse the incoming representation and map it to a model instance. Use the `#consume!` method for this.
The client must provide a `Content-Type` request header with proper MIME type to let `#consume!` know which representer to use.

```ruby
class SingersController < ApplicationController
  respond_to :json

  def create
    singer = Singer.new
    consume!(singer)

    respond_with singer
  end
end
```

For instance, if content type is set to `application/xml` the `consume!` call will roughly do the following.

```ruby
singer.
  extend(SingerRepresenter)
  from_xml(request.body)
```

So, `#consume!` helps you figuring out the representer module and reading the incoming document. Just like Rails, depending on the registered MIME type for `Content-type` it picks the deserialize method (e.g. `from_json` vs. `from_xml`)

It is important to provide a known content type in the request. If it is missing or not supported by the responder
`#consume!` will raise an exception `Roar::Rails::ControllerAdditions::UnsupportedMediaType`. Unless you rescue the exception the action will stop and respond with HTTP status `406 Unsupported Media Type`.

Note that `#consume!` respects settings from `#represents`. It uses the same mechanics known from `#respond_with` to choose a representer.

```ruby
consume!(singer, :represent_with => MusicianRepresenter)
```

## Using Decorators

If you prefer roar's decorator approach over extend, just go for it. roar-rails will figure out automatically which represent strategy to use. Be sure to use roar >= 0.11.17.

```ruby
class SingerRepresenter < Roar::Decorator
  include Roar::JSON
  include Roar::Hypermedia

  property :name

  link :self do
    singer_url(represented)
  end
end
```

In decorators' link blocks you currently have to use `represented` to get the actual represented model (this is `self` in module representers).

### Pagination with Decorators
Roar-rails pagination works with either [will_paginate](https://github.com/mislav/will_paginate) or [Kaminari](https://github.com/amatsuda/kaminari). It is based on Nick Sutterer's [blog post](http://nicksda.apotomo.de/2012/05/ruby-on-rest-6-pagination-with-roar/). You must install one of these for roar-rails pagination to work. In your Gemfile, include either

```ruby
gem 'will_paginate'
```
or

```ruby
gem 'kaminari'
```

To use roar-rails pagination, include `Roar::Rails::PageRepresenter` in the paginated Representer and define `page_url`.

```ruby
class VenuesRepresenter < Roar::Decorator
  include Roar::Representer::JSON
  include Roar::Representer::Feature::Hypermedia
  include Roar::Rails::PageRepresenter

  collection :venues, :exec_context => :decorator, :decorator => VenueRepresenter

  def venues
    represented
  end

  def page_url(args)
    venues_url args # Using Rails URL helpers
  end
end

class VenueRepresenter < Roar::Decorator
  include Roar::Representer::JSON
  include Roar::Representer::Feature::Hypermedia

  property :name
end

# Using will_paginate
class VenuesController < ActionController::Base
  include Roar::Rails::ControllerAdditions
  represents :json, Venue

  def index
    @venues = Venue.paginate(:page => params[:page], :per_page => params[:per_page])

    respond_with @venues
  end
end

# Using Kaminari
class VenuesController < ActionController::Base
  include Roar::Rails::ControllerAdditions
  represents :json, Venue

  def index
    @venues = Venue.page(params[:page]).per(params[:per_page])

    respond_with @venues
  end
end
```

**GET** `/venues?page=2&per_page=1` would give you a response similar to:

```json
{
  "total_entries": 3,
  "links": [
    {
      "rel": "self",
      "href": "http://roar.apotomo.de/venues?page=2&per_page=1"
    },
    {
      "rel": "next",
      "href": "http://roar.apotomo.de/venues?page=3&per_page=1"
    },
    {
      "rel": "previous",
      "href": "http://roar.apotomo.de/venues?page=1&per_page=1"
    }
  ],
  "venues": [
    {
      "name": "The Gorge"
    }
  ]
}
```

You can define additional pagination properties, such as `per_page`, that are provided by [will_paginate](https://github.com/mislav/will_paginate) and [Kaminari](https://github.com/amatsuda/kaminari) in your response by defining the property in your paginated representer.

```ruby
class VenuesRepresenter < Roar::Decorator
  include Roar::Representer::JSON
  include Roar::Representer::Feature::Hypermedia
  include Roar::Rails::PageRepresenter

  property :per_page

  collection :venues, :exec_context => :decorator, :decorator => VenueRepresenter

  def venues
    represented
  end

  def page_url(args)
    venues_url args
  end
end

```

## Passing Options

Both rendering and consuming support passing user options to the representer.

With `#respond_with`, any additional options will be passed to `to_json` (or whatever format you're using).

```ruby
respond_with @singer, :current_user => current_user
```

Same goes with `#consume!`, passing options to `from_json`.

```ruby
consume! Singer.new, :current_user => current_user
```

Note: If you pass in options to a representer, you must process them youself.  For rendering, use `:getter` in the representer.

```ruby
property :username, getter: lambda { |args| args[:current_user].name }
```

That'll render the `current_user`'s name as the `username` property.

More docs about passing and processing option can be found [here](https://github.com/apotonick/representable/#passing-options).


## URL Helpers

Any URL helpers from the Rails app are automatically available in representers.

```ruby
module FruitRepresenter
  include Roar::JSON
  include Roar::Hypermedia

  link :self do
    fruit_url self
  end
end
```
To get the hyperlinks up and running, please make sure to set the right _host name_ in your environment files (config/environments):

```ruby
config.representer.default_url_options = {:host => "127.0.0.1:3000"}
```

Attention: If you are using representers from a gem your Rails URL helpers might not work in these modules. This is due to a [loading order problem](https://groups.google.com/forum/?fromgroups#!topic/rubyonrails-core/5tG5unZ8jDQ) in Rails. As a workaround, don't require the representers in the gem but load them as late as possible, usually it works when you require in the controller. We are working on fixing that problem.

## Representing Formats Exclusively

By default, roar-rails will extend/decorate any model passed to `respond_with` for any request format. When adding roar-rails to a legacy project, you might want to restrict roar-rails' representing and fall back to the old behavior for certain formats. This can be configured both globally and on a per action basis.

To restrict representing globally to a particular format you can set the `config.representer.represented_formats` in your environment's configuration to an array of formats.  For example the following will only represent hal and json requests.

```ruby
config.representer.represented_formats = [:hal, :json]
```

The global configuration (or lack thereof) can be overridden by supplying the `:represented_formats` array when calling `respond_with`.  The following will only represent `@resource` for the hal format in the `show` action. For any other format, it will expose the resource using Rails' old behavior.


```ruby
class MyController < ApplicationController
  def show
    ...
    respond_with @resource, :represented_formats => [:hal]
  end
end
```

You can entirely suppress roar-rails in `respond_with` by passing in an empty array.

```ruby
class MyController < ApplicationController
  def show
    ...
    respond_with @resource, :represented_formats => []
  end
end
```

## Testing

## Autoloading

Put your representers in `app/representers` and they will be autoloaded by Rails. Also, frequently used modules as media representers and features don't need to be required manually.

## JSON-API

In a JSON-API environment, only one representer is written for both singular and collection resources. However, you have to configure `represents` accordingly so it knows what representer to use for collections.

```ruby
class SongsController < ApplicationController
  represents :json_api, entity: SongRepresenter, collection: SongRepresenter.for_collection
```


## Rails 4.1+ and HAL/JSON-API

**Note**: this is a temporary work-around, we're trying to fix that in Rails/roar-rails itself [May 2014].

Rails 4.1 and up expects you to manually register a global HAL or JSON-API renderer, or `respond_with` will throw an `ActionController::MissingRenderer` exception.

One fix is to add this to `config/initializers/mime_types.rb` right below `Mime::Type.register 'application/hal+json', :hal`:

```ruby
ActionController::Renderers.add :hal do |obj, options|
  self.content_type ||= Mime[:hal]
  obj
end
```

Similarly, for JSON-API, below `Mime::Type.register 'application/vnd.api+json', :json_api` add:


```ruby
ActionController::Renderers.add :json_api do |obj, options|
  self.content_type ||= Mime[:json_api]
  obj
end
```
## Contributors

* [Railslove](http://www.railslove.de) and especially Michael Bumann [bumi] have heavily supported development of roar-rails ("resource :singers").

[responders]: https://github.com/plataformatec/responders

## License

Roar-rails is released under the [MIT License](http://www.opensource.org/licenses/MIT).


CHANGES.markdown
## Next

## 1.2.0

* Relax version requirement for Roar

## 1.1.0

**Roar 1.1 (and Representable 3.0) support** (see [Roar CHANGES](https://github.com/trailblazer/roar/blob/master/CHANGES.markdown#110))

* Require Roar >= 1.1.0.
* Improved support for JSON API (now requires separate [roar-jsonapi](https://github.com/trailblazer/roar-jsonapi) gem).
* Fix Uber version requirement.

## 1.0.2

* Make Roar version requirement stricter: restrict to 1.0.x series.
* Initial support for Rails 5.

## 1.0.1

* Added a dependency on `responders` gem for Rails 4.2 support
* `#render` from `ControllerAdditions::Render` now accepts a status code
* Added support for JSON-API.

## 1.0.0

* Requires Roar >= 1.0.0.
* In `#consume!` roar-rails now finds the correct representer for the `Content-type:` header. In former version the representer name was inferred using the `Accept:` header, which was totally wrong. Thanks to @pgaertig for fixing that.

## 0.1.6

* Added `ControllerAdditions::Render` to support `#render` in controller actions with representers.
* Roar-rails now works properly with rails-api.
* `Roar::Rails::Responder` is a class now.
* Moved `RepresenterComputer` to `Formats`.

## 0.1.5

* Maintenance release, using Uber now and get rid of deprecation warning.

## 0.1.4

* Added a generator to create modules from the command-line. Thanks to Guilherme Cavalcanti <guiocavalcanti> for working on this.
* We always use `request.body.read` now, however, we call `rewind` beforehand.

## 0.1.3

* Fix a bug where `TestCase` wouldn't work in 3.2.

## 0.1.2

* Introducing a check where `request.body.read` is called (as usual) if `request.body.string` isn't present. This is due to some faulty behaviour in Rails and some web servers that don't provide a properly working `#read`, as it's the Rack rule. Should finally fix https://github.com/apotonick/roar-rails/issues/18 and friends.
* `TestCase#process` no longer triggers a deprecation warning for Rails > 3.1.
* Now allows using roar >= 0.11.

## 0.1.1

* Added the `represented_formats: [...]` option to be passed to `#respond_with` for either suppressing roar-rails from extending/decorating models when rendering or to fine-tune so this will only happen on white-listed formats as `:hal`. This can also be set globally using `config.representer.represented_formats`.

## 0.1.0

* `ActiveRecord::Relation` is now detected as a collection and the appropriate representer should be found.
* Entity (singular) representers are now correctly inferred even if you only specified `collection:` in `::represents`. That works by querying the model.
* Entity representers are now *namespaced* when guessed (i.e., when you didn't specify them explicitly in `::represents`) as it works with collection representers already. If you have a namespaced controller and it suddenly doesn't find its entity representer anymore, either namespace the representer or specify its name in `::represents`.

## 0.0.14

* Moved logic to infer representer names from `ControllerAdditions` to `RepresenterComputer` class.
* Representer names passed to `::represents` are now constantized at runtime where they are actually needed, only. This fixes a bug where you were required to provide a `SingersRepresenter` (for collections) everywhere even when you just want to represent singular resources.
* You can now pass strings to `::represents` as representer names.

## 0.0.13

* Allow passing user options to both `#respond_with` and `#consume!`.
* Fixing `#consume!` with decorators.

## 0.0.12

* Bumping representable to 1.4 which allows us using both extend and decorating representers.

## 0.0.11

* Back to `request.body.read` in `#consume!` (again). If anyone is having problems with empty incoming strings this is a Rails issue - update ;-)

## 0.0.10

* Empty resources now work properly in Rails 3.0 and 3.1.

## 0.0.9

* Changed `request.body.string` to `read` to make it work with Unicorn.

## 0.0.8

* Added `#represents` to configure consuming and rendering on controller class layer. This also calls `respond_to`.

## 0.0.7

* Introduce `:represent_with` and `:represent_items_with` for `#respond_with`. In turn, deprecate the old behaviour since it will change in 1.0.

## 0.0.6

* Make roar-rails rely on roar 0.10 and bigger.

## 0.0.4

* Added `#consume!` to infer representer name and parse the incoming document for a model.

## 0.0.3

* Works with Rails 3.0 now, too. Fixed the `mounted_helpers` import.


lib/roar-rails.rb
require 'roar/rails/version'
require 'roar/representer'
require 'roar/decorator'
require 'roar/rails/railtie'

module Roar
  autoload('XML', 'roar/xml')
  autoload('JSON', 'roar/json')

  module JSON
    autoload('HAL', 'roar/rails/hal')
    autoload('JSONAPI', 'roar/rails/json_api')
  end

  autoload('Hypermedia', 'roar/hypermedia')
end

module Roar
  module Rails
    def self.rails_version
      Gem::Version.new([ActionPack::VERSION::MAJOR, ActionPack::VERSION::MINOR].join('.'))
    end

    case rails_version
    when Gem::Version.new(3.0)
      require 'roar/rails/rails3_0_strategy'
    when Gem::Version.new(3.1)
      require 'roar/rails/rails3_1_strategy'
    when Gem::Version.new(3.2)
      require 'roar/rails/rails3_2_strategy'
    when Gem::Version.new(4.0), Gem::Version.new(4.1)
      require 'roar/rails/rails4_0_strategy'
    when Gem::Version.new(4.2)
      require 'roar/rails/rails4_2_strategy'
    when Gem::Version.new(5.0)
      require 'roar/rails/rails5_0_strategy'
    when Gem::Version.new('5.1')
      require 'roar/rails/rails5_0_strategy'
    when Gem::Version.new('5.2')
      require 'roar/rails/rails5_0_strategy'
    when Gem::Version.new('7.0')
      require 'roar/rails/rails7_0_strategy'
    else
      # fallback to 4.0 strategy
      require 'roar/rails/rails4_0_strategy'
    end

    autoload('TestCase', 'roar/rails/test_case')
  end
end

require 'roar/rails/controller_additions'
